pkg/phlaredb/deduplicating_slice.go

```diff
@@ -110,17 +109,14 @@ func (s *deduplicatingSlice[M, K, H, P]) maxRowsPerRowGroup() int {
 }
 
 func (s *deduplicatingSlice[M, K, H, P]) Flush(ctx context.Context) (numRows uint64, numRowGroups uint64, err error) {
-	s.lock.Lock()
-	defer s.lock.Unlock()
-
-	// intialise buffer if not existing
-	if s.buffer == nil {
-		s.buffer = parquet.NewBuffer(
-			s.persister.Schema(),
-			parquet.SortingRowGroupConfig(s.persister.SortingColumns()),
-			parquet.ColumnBufferCapacity(s.cfg.MaxBufferRowCount),
-		)
-	}
+	s.lock.RLock()
+	defer s.lock.RUnlock()
+
+	buffer := parquet.NewBuffer(
```

Comment:
Not possible to re-use it anymore ?

Reply:
I'm afraid I failed to find where we reuse dedup slice or its parts: we use zero values at head initialization: https://github.com/grafana/phlare/blob/d3033486455ae74ff4c5938eaaa4c2ff15b39240/pkg/phlaredb/head.go#L158 

and the `Init` function does not seem to try to reuse allocated objects by any means:
https://github.com/grafana/phlare/blob/d3033486455ae74ff4c5938eaaa4c2ff15b39240/pkg/phlaredb/deduplicating_slice.go#L61-L77

The buffer itself is not used anywhere but at `Flush`.

I should say that if we want to flush the head without blocking writes to the db, we have to sacrifice some memory. Although a pool could help there, judging by the GC rate, it's very likely the objects will be removed from it before the next flush. My conclusion is, provided that I don't find this allocation particularly harmful (it is almost unnoticeable in the heap profile), performance improvement would be small to unnoticeable. Therefore I'm not very concerned about this